### ISSUE TITLE
timeout too long for initial connection failures, blocks complete fluentd startup

### ISSUE THREAD

**Issue by juliantaylor** [uken/fluent-plugin-elasticsearch#503]
#### Problem

We had an issue that an elasticsearch instance was not reachable with timeout issues. Our fluentd is currently configured to have many different match sections and only some of them are using the elasticsearch plugin. Also there are 19 different match sections using the elasticsearch plugin for the same elasticsearch endpoint that had the timeout issue.
The problem in fluentd was that no output, even the ones using different plugins, did not process any data.

The log contained many entries containing:

    Could not connect Elasticsearch or obtain version. Assuming Elasticsearch 5.

All several minutes apart.

We suspect the problem was that for each match using the plugin it tries to connect and sits several minutes in a timeout and thus blocks fluentd from starting up the other matches essentially blocking the whole system.
We haven't confirmed this yet as we did not wait for several ours to for all timeouts to elapse but we can recheck the case if you think our suspicion is incorrect.

Here the log output 
```
1340 2018-11-21 11:56:30 +0000 [info]: parsing config file is succeeded path="/etc/td-agent/td-agent.conf"
1352 2018-11-21 12:22:45 +0000 [warn]: Could not connect Elasticsearch or obtain version. Assuming Elasticsearch 5.
1353 2018-11-21 12:35:52 +0000 [warn]: Could not connect Elasticsearch or obtain version. Assuming Elasticsearch 5.
1354 2018-11-21 12:48:58 +0000 [warn]: Could not connect Elasticsearch or obtain version. Assuming Elasticsearch 5.
1355 2018-11-21 13:02:04 +0000 [warn]: Could not connect Elasticsearch or obtain version. Assuming Elasticsearch 5.
1356 2018-11-21 13:15:11 +0000 [warn]: Could not connect Elasticsearch or obtain version. Assuming Elasticsearch 5.
1357 2018-11-21 13:28:17 +0000 [warn]: Could not connect Elasticsearch or obtain version. Assuming Elasticsearch 5.
1358 2018-11-21 13:41:24 +0000 [warn]: Could not connect Elasticsearch or obtain version. Assuming Elasticsearch 5.
1359 2018-11-21 13:54:30 +0000 [warn]: Could not connect Elasticsearch or obtain version. Assuming Elasticsearch 5.
1360 2018-11-21 14:07:37 +0000 [warn]: Could not connect Elasticsearch or obtain version. Assuming Elasticsearch 5.
1361 2018-11-21 14:20:43 +0000 [warn]: Could not connect Elasticsearch or obtain version. Assuming Elasticsearch 5.
1362 2018-11-21 14:33:49 +0000 [warn]: Could not connect Elasticsearch or obtain version. Assuming Elasticsearch 5.
```

As you can see there are about 13 minutes between each log warning and no other outputs are initialized.

As for a solution I guess that might be difficult in a plugin the equal matches cannot be aggregated into a single startup check, though the timeout seems to be pretty long, a shorter timeout might help to at least not have a startup that can take hours.

#### fluentd config

The relevant fluentd configuration is, this match was repeated 19 times for different tags

```
<match kubernetes.pods.somenamspace.**>

@type elasticsearch
hosts host1:9200,host2:9200,host3:9200
scheme https
ssl_version TLSv1_2
ca_file /etc/ssl/certs/ca-certificates.crt
user someuser
password %{somepassword}
include_tag_key true
logstash_format true
logstash_dateformat %Y.%m.%d-%H:00
logstash_prefix app-logs
disable_retry_limit false
retry_limit 17
max_retry_wait 10s # default is infinite
reconnect_on_error true
</match>

```

#### Using Fluentd and ES plugin versions
Th fluentd is running in a container in kubernetes.

* OS version CoreOS 1855.4.0
* Fluentd td-agent 3.2.0 (elastic plugin 2.10.3)

```
2018-09-12 15:56:17 +0000 [info]: starting fluentd-1.2.2 pid=1 ruby="2.4.4"
2018-09-12 15:56:17 +0000 [info]: spawn command to main:  cmdline=["/opt/td-agent/embedded/bin/ruby", "-Eascii-8bit:ascii-8bit", "/usr/sbin/td-agent", "--under-supervisor"]
2018-09-12 15:56:18 +0000 [info]: gem 'fluent-plugin-elasticsearch' version '2.10.3'
2018-09-12 15:56:18 +0000 [info]: gem 'fluent-plugin-kafka' version '0.7.3'
2018-09-12 15:56:18 +0000 [info]: gem 'fluent-plugin-kubernetes_metadata_filter' version '2.1.2'
2018-09-12 15:56:18 +0000 [info]: gem 'fluent-plugin-out-http' version '0.2.0'
2018-09-12 15:56:18 +0000 [info]: gem 'fluent-plugin-prometheus' version '1.0.1'
2018-09-12 15:56:18 +0000 [info]: gem 'fluent-plugin-record-modifier' version '1.1.0'
2018-09-12 15:56:18 +0000 [info]: gem 'fluent-plugin-rewrite-tag-filter' version '2.1.0'
2018-09-12 15:56:18 +0000 [info]: gem 'fluent-plugin-s3' version '1.1.3'
2018-09-12 15:56:18 +0000 [info]: gem 'fluent-plugin-systemd' version '1.0.1'
2018-09-12 15:56:18 +0000 [info]: gem 'fluent-plugin-td' version '1.0.0'
2018-09-12 15:56:18 +0000 [info]: gem 'fluent-plugin-td-monitoring' version '0.2.4'
2018-09-12 15:56:18 +0000 [info]: gem 'fluent-plugin-webhdfs' version '1.2.3'
2018-09-12 15:56:18 +0000 [info]: gem 'fluentd' version '1.2.2'
```

* ES version 6.3.0
* searchguard version 22.3

The issue is also reproducable with td-agent 3.2.1 and elastic plugin v2.11.11

**Comment by juliantaylor** [comment]
Thanks for adding a way one can avoid this problem, I can test your change next week.

Though our configuration is also pretty bad we should probably just de-duplicate the matches to avoid too long delays.

**Comment by cosmo0920** [comment]
I've released this fix as v2.12.2.
Could you try it?
